### DISCUSSION TITLE
Insertion point in Beams elements

### DISCUSSION THREAD

**Discussion by ebrahimraeyat** [xcfem/xc#83] [Q&A]
Hi @lcpt. As you know beams on structures most of the time align from top. If we want to consider this situation in XC, it there any suggestion:
examples in steel and concrete structures in ETABS:

![image](https://user-images.githubusercontent.com/8196112/109156805-883db300-7786-11eb-8438-bfa15fe3ec22.png)

![image](https://user-images.githubusercontent.com/8196112/109157463-4a8d5a00-7787-11eb-8f64-5b801c5312d6.png)

related link:
https://wiki.csiamerica.com/display/kb/Insertion+point

**Comment by lcpt** [comment]
Hi @ebrahimraeyat 

We can mimic this behaviour by creating a "wrapping" element in C++ that deals with the necessary transformations for linear elements (elastic beams, force beam columns, displacement beam columns, etc.) and maybe for shell elements too.

If someone wants to take the lead of this development I'll be glad to provide support.

**Comment by ebrahimraeyat** [comment]
Thanks, @lcpt. I want to do this, but unfortunately I'm not familiar with c++ language. I will start to learn c++ soon.

**Comment by lcpt** [comment]
Hi @ebrahimraeyat 

Your knowledge about Python will help you a lot. If I can help, please ask.

**Comment by ebrahimraeyat** [comment]
What is the best references for learning c++? Thanks.

**Reply by lcpt** [reply]
In my opinion the best references are Mr. Stroustroup books: [The C++ programming language](https://stroustrup.com/4th.html) and [The annotated C++ reference manual](https://stroustrup.com/arm.html), but I think also that it's better to start with a more friendly book for beginners where you can start familiarizing yourself with the basic concepts. Many years ago I've used a book titled "C++ from square one" but apparently it's not printed any more. I think any "C++ for beginners" book will do. Nevertheless, it's important not to get distracted by accessory aspects like user interface design etc. Some books start to explain how to use Microsoft Foundations classes or something similar and all that is perfectly useless if you want to code for finite element analysis.

I'll take a look and try to recommend you some references.

**Reply by lcpt** [reply]
There you go:

- [A Tour of C++ (C++ In-Depth Series) 1st Edition](https://geni.us/6fnnXpc) Bjarne Stroustrup. For the brave and the impatient :)
- [C++ Primer 5th Edition](https://geni.us/Y6Udqa).
- [C++ Pocket Reference 1st Edition](https://geni.us/w3tnB9).

**Comment by ebrahimraeyat** [comment]
Thanks, @lcpt. I will start it soon.
